src/python/WMCore/ReqMgr/DataStructs/Request.py

```diff
@@ -88,31 +91,53 @@ def _replace_cloned_args(clone_args, user_args):
     for prop in user_args:
         if isinstance(user_args[prop], dict) and isinstance(clone_args.get(prop), dict):
             _replace_cloned_args(clone_args.get(prop, {}), user_args[prop])
-        else:     
+        else:
             clone_args[prop] = user_args[prop]
     return
-    
-def initialize_clone(request_args, reqmgr_db_service):
+
+
+def initialize_clone(request_args, originalArgs, argsDefinition, chainDefinition=None):
     """
-    Initialize a Clone arguments by inheriting and overwriting argument from OriginalRequest
+    Initialize arguments for a clone request by inheriting and overwriting argument
+    from OriginalRequest.
+
+    :param request_args: user-provided dictionary with override arguments
+    :param originalArgs: original arguments retrieved for the workflow being cloned
+    :param argsDefinition: arguments definition according to the workflow type being cloned
+    :param chainDefinition: a dictionary containing the chain argument definition, for
+    SteoChain and TaskChain
+    :return: dictionary with original args filtered out, as per the spec definition. And on
+     top of that, user arguments added/replaced in the dictionary.
     """
-    requests = reqmgr_db_service.getRequestByNames(request_args["OriginalRequestName"])
-    clone_args = requests.values()[0]
-    # TODO: need to validate new overwrite request_args here since it will skip the clone_args validtaiton
+    chainPattern = r'(Task|Step)\d{1,2}'
+    clone_args = {}
+    for topKey, topValue in originalArgs.iteritems():
+        if topKey in argsDefinition:
+            clone_args[topKey] = topValue
+        elif topKey.startswith(("Skim", "Step", "Task")):
+            # accept floating args from ReReco, StepChain and TaskChain
+            if re.match(chainPattern, topKey):
+                for innerKey in topValue:
```

Comment:
btw, I don't like it. There is likely a one line code that does it all, I just need to figure it out :)